File: security.ts
/**
 * security — dependency-free hardening middleware for the Jarvis brain server.
 *
 * OWASP-aligned for a localhost single-user service:
 *   • rateLimit          — in-memory token-bucket limiter, per client IP, 429 + Retry-After.
 *   • securityHeaders     — nosniff / frame-deny / referrer / CSP.
 *   • requireLoopbackHost — Host-header allow-list (anti DNS-rebinding).
 *   • validateBody        — strict allow-list schema: type, length, required,
 *                           and rejection of unexpected fields (no mass-assignment).
 *
 * All in-memory, no new packages — keeps a fresh clone booting unchanged.
 */
import type { Request, Response, NextFunction } from "express";

// ── token-bucket rate limiter ─────────────────────────────────────
interface Bucket {
  tokens: number;
  updated: number;
}

export interface RateLimitOptions {
  /** max burst */
  capacity: number;
  /** sustained refills per second once the burst is spent */
  refillPerSec: number;
  /** label used in the 429 message */
  name?: string;
}

/** Build an express middleware that token-bucket rate-limits by client IP. */
export function rateLimit(opts: RateLimitOptions) {
  const buckets = new Map<string, Bucket>();

  // Periodically drop idle buckets so memory can't grow unbounded.
  const idleMs = 60 * 60 * 1000;
  setInterval(() => {
    const cutoff = Date.now() - idleMs;
    for (const [k, b] of buckets) if (b.updated < cutoff) buckets.delete(k);
  }, idleMs).unref?.();

  return (req: Request, res: Response, next: NextFunction) => {
    const key = req.ip ?? req.socket.remoteAddress ?? "unknown";
    const now = Date.now();
    let b = buckets.get(key);
    if (!b) {
      b = { tokens: opts.capacity, updated: now };
      buckets.set(key, b);
    }
    // Refill proportional to elapsed time, capped at capacity.
    const elapsedSec = (now - b.updated) / 1000;
    b.tokens = Math.min(opts.capacity, b.tokens + elapsedSec * opts.refillPerSec);
    b.updated = now;

    if (b.tokens >= 1) {
      b.tokens -= 1;
      next();
      return;
    }
    const retryAfter = Math.max(1, Math.ceil((1 - b.tokens) / opts.refillPerSec));
    res.setHeader("Retry-After", String(retryAfter));
    res.status(429).json({ error: "Too many requests — slow down." });
  };
}

// ── security headers ──────────────────────────────────────────────
export function securityHeaders(_req: Request, res: Response, next: NextFunction) {
  res.setHeader("X-Content-Type-Options", "nosniff");
  res.setHeader("X-Frame-Options", "DENY");
  res.setHeader("Referrer-Policy", "no-referrer");
  // The chat UI uses inline <style>/<script> and fetches the same origin.
  res.setHeader(
    "Content-Security-Policy",
    "default-src 'none'; style-src 'unsafe-inline'; script-src 'unsafe-inline'; " +
      "connect-src 'self'; img-src 'self' data:; media-src 'self' blob:",
  );
  next();
}

// ── Host allow-list (anti DNS-rebinding) ──────────────────────────
const LOOPBACK = new Set(["127.0.0.1", "localhost", "::1", "[::1]"]);

export function requireLoopbackHost(req: Request, res: Response, next: NextFunction) {
  const host = (req.headers.host ?? "").split(":")[0].trim().toLowerCase();
  if (!LOOPBACK.has(host)) {
    res.status(403).json({ error: "Forbidden host." });
    return;
  }
  next();
}

// ── strict body validation ────────────────────────────────────────
export interface FieldSpec {
  type: "string" | "number" | "boolean";
  required?: boolean;
  maxLen?: number; // strings
  min?: number; // numbers
  max?: number; // numbers
  trim?: boolean; // strings
}

export type Schema = Record<string, FieldSpec>;

export class ValidationError extends Error {}

/**
 * Validate & normalise `body` against `schema`. Rejects non-objects, unexpected
 * fields, wrong types, and over-length strings. Returns a clean object holding
 * only declared fields.
 */
export function validate(body: unknown, schema: Schema): Record<string, unknown> {
  if (typeof body !== "object" || body === null || Array.isArray(body)) {
    throw new ValidationError("body must be a JSON object");
  }
  const obj = body as Record<string, unknown>;

  for (const key of Object.keys(obj)) {
    if (!(key in schema)) throw new ValidationError(`unexpected field: ${key}`);
  }

  const out: Record<string, unknown> = {};
  for (const [name, spec] of Object.entries(schema)) {
    const present = obj[name] !== undefined && obj[name] !== null;
    if (!present) {
      if (spec.required) throw new ValidationError(`missing required field: ${name}`);
      continue;
    }
    let val = obj[name];

    if (spec.type === "string") {
      if (typeof val !== "string") throw new ValidationError(`${name} must be a string`);
      if (spec.trim) val = val.trim();
      if ((val as string).length === 0 && spec.required) {
        throw new ValidationError(`${name} must not be empty`);
      }
      if (spec.maxLen !== undefined && (val as string).length > spec.maxLen) {
        throw new ValidationError(`${name} exceeds max length ${spec.maxLen}`);
      }
    } else if (spec.type === "number") {
      if (typeof val !== "number" || Number.isNaN(val)) {
        throw new ValidationError(`${name} must be a number`);
      }
      if (spec.min !== undefined && val < spec.min) throw new ValidationError(`${name} too small`);
      if (spec.max !== undefined && val > spec.max) throw new ValidationError(`${name} too large`);
    } else if (spec.type === "boolean") {
      if (typeof val !== "boolean") throw new ValidationError(`${name} must be a boolean`);
    }

    out[name] = val;
  }
  return out;
}

/** Validation middleware factory: validates req.body, replaces it with the clean object. */
export function validateBody(schema: Schema) {
  return (req: Request, res: Response, next: NextFunction) => {
    try {
      req.body = validate(req.body, schema);
      next();
    } catch (err) {
      res.status(422).json({ error: `Invalid input: ${(err as Error).message}` });
    }
  };
}

File: actions/security.py
"""
security — reusable, dependency-free hardening primitives for the local HTTP
control surfaces (content_agent_server.py, and usable by any future Python
endpoint).

Implements, per OWASP guidance for a localhost single-user control plane:
  • RateLimiter         — token-bucket limiter keyed by (IP + session token).
  • validate_schema     — strict allow-list schema validation (type / length /
                          choices / regex), rejecting unexpected fields.
  • CSRF / origin guards — per-process token + Host/Origin allow-listing to stop
                          DNS-rebinding and a malicious web page in the user's
                          browser from forging requests to 127.0.0.1.
  • constant_time_eq    — timing-safe token comparison.

Everything is in-memory and stdlib-only — no new dependencies, nothing that can
leak secrets, and safe to import from anywhere.
"""
from __future__ import annotations

import hmac
import re
import secrets
import threading
import time
from dataclasses import dataclass, field
from typing import Any, Iterable, Optional
from urllib.parse import urlparse

# ── CSRF / session token ──────────────────────────────────────────
# One random token per server process. It is injected into the served HTML so
# the same-origin page can send it back; cross-origin pages cannot read it, so
# they cannot forge state-changing requests (CSRF defence).
SESSION_TOKEN: str = secrets.token_urlsafe(32)


def constant_time_eq(a: str, b: str) -> bool:
    """Timing-safe string compare (avoids leaking token length/prefix)."""
    return hmac.compare_digest((a or "").encode(), (b or "").encode())


# ── Host / Origin allow-listing (anti DNS-rebinding & CSRF) ────────
def host_allowed(host_header: str, port: int) -> bool:
    """Accept only loopback Host headers for the bound port.

    Blocks DNS-rebinding (where a public hostname resolves to 127.0.0.1 and a
    remote page drives the local server)."""
    if not host_header:
        return False
    host = host_header.split(":", 1)[0].strip().lower()
    return host in {"127.0.0.1", "localhost", "[::1]", "::1"}


def origin_allowed(origin_or_referer: str, port: int) -> bool:
    """True when an Origin/Referer header points at our own loopback server.

    An *absent* Origin is allowed (non-browser clients / same-origin GETs);
    a *present* cross-origin one is rejected."""
    if not origin_or_referer:
        return True
    try:
        u = urlparse(origin_or_referer)
    except Exception:
        return False
    return u.hostname in {"127.0.0.1", "localhost", "::1"}


# ── Token-bucket rate limiter ─────────────────────────────────────
@dataclass
class _Bucket:
    tokens: float
    updated: float


@dataclass
class RateLimiter:
    """Thread-safe token bucket.

    capacity  — max burst.
    refill_per_sec — sustained requests/second once burst is spent.
    A request costs 1 token; when the bucket is empty the caller is told how
    long (seconds) until the next token, for a graceful ``Retry-After``.
    """
    capacity: float
    refill_per_sec: float
    _buckets: dict[str, _Bucket] = field(default_factory=dict)
    _lock: threading.Lock = field(default_factory=threading.Lock)

    def check(self, key: str) -> tuple[bool, int]:
        """Return (allowed, retry_after_seconds)."""
        now = time.monotonic()
        with self._lock:
            b = self._buckets.get(key)
            if b is None:
                b = _Bucket(tokens=self.capacity, updated=now)
                self._buckets[key] = b
            # Refill based on elapsed time, capped at capacity.
            elapsed = now - b.updated
            b.tokens = min(self.capacity, b.tokens + elapsed * self.refill_per_sec)
            b.updated = now
            if b.tokens >= 1.0:
                b.tokens -= 1.0
                return True, 0
            # Empty — seconds until one token refills.
            retry = max(1, int((1.0 - b.tokens) / self.refill_per_sec) + 1)
            return False, retry

    def prune(self, max_idle_sec: float = 3600.0) -> None:
        """Drop buckets unused for a while (prevents unbounded growth)."""
        cutoff = time.monotonic() - max_idle_sec
        with self._lock:
            for k in [k for k, v in self._buckets.items() if v.updated < cutoff]:
                self._buckets.pop(k, None)


# ── Brute-force lockout ───────────────────────────────────────────
@dataclass
class BruteForceGuard:
    """Locks out a key (typically a client IP) after too many FAILED auth attempts.

    Why this is separate from RateLimiter: the request limiter is keyed by
    (ip + token), so each wrong token gets a fresh bucket and is never throttled
    — useless against token guessing. This guard keys on the IP alone and counts
    *failures*, so repeated bad tokens trip a temporary lockout no matter which
    token is presented.

    Usage pattern (callers enforce "valid token always passes"):
        if valid_token:  guard.register_success(ip)        # clears any lockout
        else:            locked = guard.is_locked(ip); guard.register_failure(ip)
    Because a correct token bypasses the lockout, a legitimate user on the same
    loopback IP is never self-DoS'd by an attacker hammering bad tokens.
    """
    max_failures: int = 10          # failures within the window before lockout
    window_sec: float = 60.0        # sliding window for counting failures
    lockout_sec: float = 300.0      # how long a tripped lockout lasts
    _fails: dict[str, list[float]] = field(default_factory=dict)
    _locked_until: dict[str, float] = field(default_factory=dict)
    _lock: threading.Lock = field(default_factory=threading.Lock)

    def is_locked(self, key: str) -> tuple[bool, int]:
        """Return (locked, retry_after_seconds)."""
        now = time.monotonic()
        with self._lock:
            until = self._locked_until.get(key, 0.0)
            if now < until:
                return True, max(1, int(until - now) + 1)
            return False, 0

    def register_failure(self, key: str) -> None:
        now = time.monotonic()
        with self._lock:
            arr = [t for t in self._fails.get(key, []) if t > now - self.window_sec]
            arr.append(now)
            if len(arr) >= self.max_failures:
                self._locked_until[key] = now + self.lockout_sec
                arr = []                      # reset window once locked
            self._fails[key] = arr

    def register_success(self, key: str) -> None:
        """A valid auth clears the recent FAILURE history (instant recovery for a
        legitimate client) but deliberately does NOT lift an active lockout: a
        valid token already bypasses the lockout check, so on a shared loopback
        IP a legit poll must not un-lock an attacker mid-spree. An engaged
        lockout therefore always runs its full ``lockout_sec``."""
        with self._lock:
            self._fails.pop(key, None)


# ── Strict schema validation ──────────────────────────────────────
class ValidationError(ValueError):
    """Raised when user input fails the allow-list schema."""


@dataclass
class Field:
    """A single field spec. Unknown fields in the payload are always rejected."""
    type: type                      # str | int | bool | list
    required: bool = False
    max_len: Optional[int] = None   # for str / list
    choices: Optional[Iterable[Any]] = None
    item_choices: Optional[Iterable[Any]] = None   # for list[str]
    regex: Optional[str] = None
    min_val: Optional[int] = None
    max_val: Optional[int] = None
    default: Any = None


def validate_schema(payload: Any, schema: dict[str, Field]) -> dict[str, Any]:
    """Validate & normalise ``payload`` against ``schema``.

    - rejects non-dict payloads
    - rejects any field not declared in the schema (no mass-assignment)
    - enforces type, length, numeric bounds, choices and regex
    Returns a clean dict containing only declared fields. Raises ValidationError.
    """
    if not isinstance(payload, dict):
        raise ValidationError("body must be a JSON object")

    unexpected = set(payload) - set(schema)
    if unexpected:
        raise ValidationError(f"unexpected field(s): {', '.join(sorted(unexpected))}")

    out: dict[str, Any] = {}
    for name, spec in schema.items():
        if name not in payload or payload[name] is None:
            if spec.required:
                raise ValidationError(f"missing required field: {name}")
            if spec.default is not None:
                out[name] = spec.default
            continue

        val = payload[name]

        # bool must be checked before int (bool is a subclass of int).
        if spec.type is int and isinstance(val, bool):
            raise ValidationError(f"{name} must be an integer")
        if not isinstance(val, spec.type):
            raise ValidationError(f"{name} must be {spec.type.__name__}")

        if spec.type is str:
            if spec.max_len is not None and len(val) > spec.max_len:
                raise ValidationError(f"{name} exceeds max length {spec.max_len}")
            if spec.regex is not None and not re.fullmatch(spec.regex, val):
                raise ValidationError(f"{name} has an invalid format")
            if spec.choices is not None and val not in spec.choices:
                raise ValidationError(f"{name} must be one of {list(spec.choices)}")

        elif spec.type is int:
            if spec.min_val is not None and val < spec.min_val:
                raise ValidationError(f"{name} must be >= {spec.min_val}")
            if spec.max_val is not None and val > spec.max_val:
                raise ValidationError(f"{name} must be <= {spec.max_val}")

        elif spec.type is list:
            if spec.max_len is not None and len(val) > spec.max_len:
                raise ValidationError(f"{name} has too many items (max {spec.max_len})")
            if spec.item_choices is not None:
                for item in val:
                    if item not in spec.item_choices:
                        raise ValidationError(
                            f"{name} contains an invalid value: {item!r}")

        out[name] = val
    return out


def sanitize_text(s: str, max_len: int = 500) -> str:
    """Trim, cap length, and strip control characters (keeps \\n and \\t)."""
    s = (s or "")[: max_len + 1]
    s = "".join(ch for ch in s if ch in ("\n", "\t") or ord(ch) >= 0x20)
    return s.strip()[:max_len]
